addons/payment_paypal/controllers/main.py

```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
         retrieved and with it the transaction which will be immediately post-processed.
         """
         _logger.info("handling redirection from PayPal with data:\n%s", pprint.pformat(pdt_data))
-        if not pdt_data:  # The customer has canceled or paid then clicked on "Return to Merchant"
-            pass  # Redirect them to the status page to browse the (currently) draft transaction
-        else:
             # Check the origin of the notification
-            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
-                'paypal', pdt_data
-            )
-            try:
-                notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
-            except Forbidden:
-                _logger.exception("could not verify the origin of the PDT; discarding it")
-            else:
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+            'paypal', pdt_data
+        )
+        try:
+            notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
+        except Forbidden:
+            _logger.exception("could not verify the origin of the PDT; discarding it")
+        else:
                 # Handle the notification data
-                tx_sudo._handle_notification_data('paypal', notification_data)
+            tx_sudo._handle_notification_data('paypal', notification_data)
+
+        return request.redirect('/payment/status')
+
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancellation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        tx_sudo = request.env['payment.transaction'].sudo().search([
+            ('reference', '=', transaction_reference),
+            ('provider_code', '=', 'paypal'),
+        ])
+        notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo=tx_sudo)
+        tx_sudo._handle_notification_data('paypal', notification_data)
```

Comment:
@Feyensv pointed out to me that we need to check the provenance of the request to avoid people canceling transactions just by making a request to this endpoint. 

`_verify_pdt_notification_origin` seemed to be a good idea except that Paypal doesn't send anything back to Odoo in that case so we can't verify anything. To check the origin of the request, you should use an access token that we can check in this controller. There is some method for that in `odoo/addons/payment/utils.py`. Have a look at `generate_access_token` and `check_access_token`. 

Since we need something to be sure that the request comes from this database, we'll use the `database_uuid` to generate the access token. That will guarantee that the request is not forged and comes from this database :)

Please find one of my suggestions below and the other in `addons/payment_paypal/models/payment_transaction.py`. 

```suggestion
        database_uuid = request.env['ir.config_parameter'].sudo().get_param('database.uuid')
        if not payment_utils.check_access_token(access_token, database_uuid, transaction_reference):
            raise Forbidden()
        tx_sudo._handle_notification_data('paypal', {})```

Reply:
I updated my suggestion following our discussion on discord. There is no need to use the `database_uuid`. Sorry for the confusion